File: server/support_test.go
package main

import (
	"math"
	"testing"
)

func TestSupportAggression(t *testing.T) {
	s := initial()
	s.logicalView = [2]int{0, 484}
	place := func(i, x, y int) {
		s.Players[i].X = float64(576-y) * (22.4 / 576)
		s.Players[i].Z = float64(x-320) * (22.4 / 576)
	}
	place(3, 120, 576)
	for j := 9; j < 18; j++ {
		place(j, 600, 1000)
	}
	place(9, 160, 576)
	place(10, 180, 576)
	var distances [18]int
	for j := range distances {
		distances[j] = 999
	}
	distances[9], distances[10] = 100, 100
	check := func(random, want int) {
		t.Helper()
		x, z, ok := s.aggressionTarget(3, &distances, random)
		if want < 0 {
			if ok {
				t.Fatal("unexpected aggressive target")
			}
			return
		}
		if !ok || math.Abs(x-s.Players[want].X) > 1e-9 || math.Abs(z-s.Players[want].Z) > 1e-9 {
			t.Fatalf("target %v,%v,%v, want player %d", x, z, ok, want)
		}
	}
	check(49, 9)
	check(50, -1)
	distances[9] = 101
	check(49, 10)
	distances[9], distances[10] = 200, 200
	check(49, -1)
	distances[9] = 100
	s.Players[9].Stun = 1
	check(49, -1)
	s.Players[9].Stun = 0
	place(9, 213, 576)
	check(49, 9)
	s.Players[9].moveZ = (22.4 / 576) * 25
	check(49, -1)
	s.Players[9].moveZ = 0
	s.logicalView = [2]int{320, 484}
	check(49, -1)
}

func TestSupportTargets(t *testing.T) {
	s := initial()
	s.Controlled[0] = 8
	place := func(i, x, y int) {
		s.Players[i].X = float64(576-y) * (22.4 / 576)
		s.Players[i].Z = float64(x-320) * (22.4 / 576)
	}
	check := func(i, x, y int) {
		t.Helper()
		a, b := s.supportTarget(i)
		got := [2]int{int(math.Round(b/(22.4/576) + 320)), int(math.Round(576 - a/(22.4/576)))}
		if got != [2]int{x, y} {
			t.Fatalf("player %d target %v, want %d,%d", i, got, x, y)
		}
	}
	check(1, 248, 1032)
	place(8, 80, 80)
	check(1, 248, 1032)
	place(8, 320, 200)
	check(3, 136, 384)
	place(8, 120, 800)
	check(3, 120, 768)
	place(8, 160, 96)
	check(6, 213, 208)
	s.Period = 2
	place(8, 479, 1055)
	check(6, 426, 943)
}

File: src/support.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial } from "./game.js";
import { supportTarget, aggressionTarget } from "./support.js";
const unit = 22.4 / 576;
const target = (s, i) => {
  const [x, z] = supportTarget(s, i);
  return [Math.round(z / unit + 320), Math.round(576 - x / unit)];
};
const place = (p, x, y) =>
  Object.assign(p, { x: (576 - y) * unit, z: (x - 320) * unit });

test("support aggression uses strict random and distance thresholds, first tie, and predicted role bounds", () => {
  const s = initial();
  s.logicalView = [0, 484];
  place(s.players[3], 120, 576);
  for (let j = 9; j < 18; j++) place(s.players[j], 600, 1000);
  place(s.players[9], 160, 576);
  place(s.players[10], 180, 576);
  const distances = Array(18).fill(999);
  distances[9] = distances[10] = 100;
  const target = (random = 49) => aggressionTarget(s, 3, distances, random);
  const xy = (j) => [s.players[j].x, s.players[j].z];
  assert.deepEqual(target(), xy(9));
  assert.equal(target(50), null);
  distances[9] = 101;
  assert.deepEqual(target(), xy(10));
  distances[9] = distances[10] = 200;
  assert.equal(target(), null);
  distances[9] = 100;
  s.players[9].stun = 1;
  assert.equal(target(), null);
  s.players[9].stun = 0;
  place(s.players[9], 213, 576);
  assert.deepEqual(target(), xy(9));
  s.players[9].moveZ = unit * 25;
  assert.equal(target(), null, "predicted position leaves the inclusive role boundary");
  s.players[9].moveZ = 0;
  s.logicalView = [320, 484];
  assert.equal(target(), null, "offscreen players retain positional support");
});
test("defender supports goalkeeper when a forward is selected", () => {
  const s = initial();
  s.controlled[0] = 8;
  assert.deepEqual(target(s, 1), [248, 1032]);
  place(s.players[8], 80, 80);
  assert.deepEqual(target(s, 1), [248, 1032]);
});
test("midfielder bends support target back into role zone", () => {
  const s = initial();
  s.controlled[0] = 8;
  place(s.players[8], 320, 200);
  assert.deepEqual(target(s, 3), [136, 384]);
  place(s.players[8], 120, 800);
  assert.deepEqual(target(s, 3), [120, 768]);
});
test("wing support lookup and halftime reflection", () => {
  const s = initial();
  s.controlled[0] = 8;
  place(s.players[8], 160, 96);
  assert.deepEqual(target(s, 6), [213, 208]);
  s.period = 2;
  place(s.players[8], 479, 1055);
  assert.deepEqual(target(s, 6), [426, 943]);
});
